File: swift/BeerSong.swift
class BeerSong {
    let amount: Int
    
    init(numberOfBeerBottles n: Int) {
        amount = n
    }
    
    func generateVersesOfBeerSong() -> String {
        return stride(from: amount, through: 0, by: -1).map { x in
            guard x > 0 else { return "No more bottles of beer on the wall, no more bottles of "
                + "beer.\nGo to the store and buy some more, 99 bottles of beer on the wall." }
            
            let firstPhrase  = "\(x) " + (x == 1 ? "bottle" : "bottles")
            let secondPhrase = x == 1 ? "no more bottles" : "\(x - 1) " + (x == 2 ? "bottle" : "bottles")
            
            return "\(firstPhrase) of beer on the wall, \(firstPhrase) of beer.\nTake one down and "
                + "pass it around, \(secondPhrase) of beer on the wall.\n\n"
            }.joined()
    }
}

File: clojure/beer_song.clj
(ns beer-song
  (:require [clojure.string :as string]))

(defn verse
  "Returns the nth verse of the song."
  [n]
  (let [bottles (str n (if (== n 1) " bottle" " bottles"))
        removed (if (== n 1) "it" "one")
        bottles-left (condp == n 1 "no more bottles" 2 "1 bottle" (str (- n 1) " bottles"))]
    (if (== n 0)
      (str "No more bottles of beer on the wall, no more bottles of beer.\n"
           "Go to the store and buy some more, 99 bottles of beer on the wall.\n")
      (str bottles " of beer on the wall, " bottles " of beer.\n"
           "Take " removed " down and pass it around, " bottles-left " of beer on the wall.\n"))))

(defn sing
  "Given a start and an optional end, returns all verses in this interval. If
  end is not given, the whole song from start is sung."
  ([start] (sing start 0))
  ([start end] (string/join "\n" (map verse (range start (- end 1) -1)))))
